src/CodeGen_OpenCL_Dev.cpp

```diff
@@ -858,8 +858,50 @@ void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const Shuffle *op) {
             }
             stream << ");\n";
         }
+    } else if (op->is_extract_element()) {
+        // OpenCL requires using .s<n> format for extracting an element
+      ostringstream rhs;
+      rhs << print_expr(op->vectors[0]);
+      rhs << ".s" << op->indices[0];
+      print_assignment(op->type, rhs.str());
+    } else if (op->type.is_scalar()) {
+        CodeGen_C::visit(op);
     } else {
-        internal_error << "Shuffle not implemented.\n";
+        internal_assert(!op->vectors.empty());
```

Comment:
This logic seems to be repeated three times. Is it really so different across the GPU backends that none of it can usefully be in the base class?

Reply:
I considered adding a base class for the various `CodeGen_<x>::CodeGen_<x>_C` classes.  Right now each of them is derived from `CodeGen_C`.  However, I decided not to do that because it felt like a larger refactoring was necessary-- we probably could consolidate a lot of the logic into a superclass, since HLSL/OpenCL C/Metal have a lot of overlap.

I'm happy to this specific functionality into a superclass if you think that's a good idea.

Reply:
I was wondering if it could go into codegen_c, but I see that the shuffle handling there relies on a helper class.

I think it's time to add a CodeGen_GPU_C intermediate class and we can slowly move stuff into it over time.